lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/IO/Resource.hs

```diff
@@ -125,11 +131,25 @@ onResource
 onResource action resource = do
     eContent <- readTVarIO $ content resource
     case eContent of
-        NotInitialized -> pure $ Left ErrNotInitialized
-        Initializing -> pure $ Left ErrStillInitializing
-        Initialized a -> Right <$> action a
+        Closed -> pure $ Left ErrNotInitialized
+        Opening -> pure $ Left ErrStillInitializing
+        Open a -> Right <$> action a
+        Vanished e -> pure $ Left $ ErrVanished e
+        FailedToOpen e -> pure $ Left $ ErrFailedToInitialize e
+        Closing -> pure $ Left ErrClosing
+
+closeResource :: Resource e a -> IO (Either (ErrResourceMissing e) ())
+closeResource resource = do
+    status <- readStatus resource
+    case status of
```

Comment:
Reading the `TVar` and changing it needs to happen in a single `atomically` block, otherwise there is a small race condition.